Vert.x Core 3.x Plugin

BFergerson opened the pull request:
- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues: #2341
___
### New feature or improvement

Basic Vert.x Core 3.x support. This includes support for the event bus and HTTP clients and servers.

wu-sheng commented:
I am flighting to SF today. 😄

BFergerson commented:
@wu-sheng, luck you :P. That's okay though. I'm sure this isn't ready yet. I wanted to ask you if I was doing the `EntrySpan`/`ExitSpan` pattern correctly for the event bus.

Basically, when you send a message through the event bus it could be going to another computer or it could be going to a consumer on the same computer. So there is a chance that it could be a message which is sent on the same thread, cross-thread, or cross process. My question is:
 - should I be detecting if the message was same thread, cross thread, or cross-process and build the appropriate span based on that?

Should I have logic like:
```
If it was sent cross-thread:
 - Ignore ContextCarrier and work with ContextSnapshot

If it was sent cross-process
 - Use ContextCarrier
```

wu-sheng commented:
```
If it was sent cross-thread:
 - Ignore ContextCarrier and work with ContextSnapshot

If it was sent cross-process
 - Use ContextCarrier
```

This is ideal status, but the more complex scenario is for span. If you call exit span, then the target ip would be required, and shown in topology if no entry span created. Also I doubt you would know when to create exit span or local span for in-process.

> Currently sending a message always creates an ExitSpan and sends the ContextCarrier and receiving a message always creates a EntrySpan using the ContextCarrier sent.

This could be an easier way. But I have a question,  if you don't know whether this event goes out-of-process or not, how set you set peer?

BFergerson commented:
@wu-sheng, why do you think it will be hard to tell whether or not to make an exit span vs a local span? I'm able to detect whether the message is going over the wire or being routed locally pretty easily. If it's going over the wire I use `ContextCarrier` and if it's not I use `ContextSnapshot`.

I'm a bit confused on how I should be handling the local routing scenario. For an example scenario, let's say I have a local message flow of: `A` -> `B` -> `C` (where `A` sends a message to `B` and so on) and that in turn triggers a response flow of: `C` -> `B` -> `A`  (where `C` responds to `B` and so on).

In this scenario, do I create any entry or exit spans or are they all supposed to be local? Your question about what I set the peer to set me down this path. Technically there wouldn't be a peer here because it's just communicating with itself. So should I just be creating a local spans?

If the answer is yes, would the process be:
 - create async local span when `A` calls `B`
 - create local span when `B` receives message from `A` (call continued with `A` context)
 - create async local span when `B` calls `C`
 - create local span when `C` receives message from `B` (call continued with `B` context)
 - close async local span when `C` responds to `B`
 - close async local span when `B` responds to `A`

I think I've got the over the wire scenario covered. It's the simple:
 - create exit span when sending messages
 - create entry span when receiving messages

wu-sheng commented:
> I'm able to detect whether the message is going over the wire or being routed locally pretty easily

If you can detect that, so you should do as you described above. 

Just remind
1. In real local, ContextSnapshot is required between AsyncA and ReceiverB
1. In `over the wire scenario`, you also need async span. 

My original questions are mostly about when you can't know it is local or remote when span is created. But look like you know. So, you could ignore my questions.

wu-sheng commented:
What is the status of this PR?

BFergerson commented:
@wu-sheng, pushed the remaining work.

Would you like me to begin working on the integration test or do you want to review/discuss the code first?

wu-sheng commented:
I think you could work on tests first. If the process is as same as we discussed, I think in theory, should not have a big problem. 

@ascrutae please begin to review the codes.

BFergerson commented:
@wu-sheng, is it necessary to open a local span when you're just closing an async local span?

Without opening another local I have this:
 - Exit span from process A
 - Entry span from process B

If I open a local span when I'm closing the async local spans, I end up with this:
 - Exit span from process A
 - Entry span from process B
 - Local span from process B to close Entry span
 - Local span from process A to close Exit span

Those local spans end up looking like:
```json
        {
          "segmentId": "1.215.15537264330420004",
          "spans": [
            {
              "operationName": "__vertx.reply.4b5d4eb7-5323-4532-aa03-4a6cec875f23",
              "operationId": "0",
              "parentSpanId": "-1",
              "spanId": "0",
              "spanLayer": "RPCFramework",
              "refs": [
                {
                  "parentEndpointId": 0,
                  "parentEndpoint": "cluster-message-receiver",
                  "networkAddressId": 0,
                  "entryEndpointId": 0,
                  "refType": "CrossThread",
                  "parentSpanId": 0,
                  "parentTraceSegmentId": "1.215.15537264330100002",
                  "parentServiceInstanceId": 1,
                  "networkAddress": "",
                  "entryEndpoint": "cluster-message-receiver",
                  "entryServiceInstanceId": 1
                }
              ],
              "startTime": "1553726433042",
              "endTime": "1553726433052",
              "componentId": "59",
              "componentName": "",
              "isError": false,
              "spanType": "Local",
              "peer": "",
              "peerId": "0"
            }
          ]
        }
```




should operation names be static or are they allowed to be dynamic? In the SkyWalking integration tests I can't find anyone who is matching on a dynamic endpoint name. The way I currently have it coded is like so:

### Send a local message
```json
{
          "segmentId": "1.213.15536967576930002",
          "spans": [
            {
              "operationName": "local-message-receiver",
              "operationId": "0",
              "parentSpanId": "-1",
              "spanId": "0",
              "spanLayer": "RPCFramework",
              "startTime": "1553696757693",
              "endTime": "1553696757715",
              "componentId": "59",
              "componentName": "",
              "isError": false,
              "spanType": "Local",
              "peer": "",
              "peerId": "0"
            }
          ]
        }
```

### Receive a local message
```json
{
              "operationName": "#local-message-receiver",
              "operationId": "0",
              "parentSpanId": "-1",
              "spanId": "0",
              "spanLayer": "RPCFramework",
              "refs": [
                {
                  "parentEndpointId": 0,
                  "parentEndpoint": "local-message-receiver",
                  "networkAddressId": 0,
                  "entryEndpointId": 0,
                  "refType": "CrossThread",
                  "parentSpanId": 0,
                  "parentTraceSegmentId": "1.213.15536967576930002",
                  "parentServiceInstanceId": 1,
                  "networkAddress": "",
                  "entryEndpoint": "local-message-receiver",
                  "entryServiceInstanceId": 1
                }
              ],
              "startTime": "1553696757696",
              "endTime": "1553696757698",
              "componentId": "59",
              "componentName": "",
              "isError": false,
              "spanType": "Local",
              "peer": "",
              "peerId": "0"
            }
          ]
        }
```

I think they always end up being in their own segment.


I believe you said before that I needed to open local spans when receiving messages because during that process the trace context could be picked up by other plugins. Does that still apply with async spans though? The point at which I'm closing the span is when the message is being responded to. I can't think of something that would need to pick up the context at that point.

If I do need to create the local spans when I'm closing async local spans, what would you recommend for naming them? When I send a message to `X` and I'm waiting for a response, Vert.x generates `__vertx.reply` so it can route back. If I need those local spans can I keep the operation name like the JSON above? I didn't see any other integrations tests with dynamic operation names so I imagine not.

wu-sheng commented:
> is it necessary to open a local span when you're just closing an async local span?

Anything else you will or plan to do when/after close async local span? If no, no need. Local span only makes sense when you want to do more thing, and try to build the link by ref(ContextSnapshot)

wu-sheng commented:
@BFergerson That test looks good. Let's wait @ascrutae review the codes. Then would we just support one certain vert.x version? Or more?

BFergerson commented:
@wu-sheng, I'm surprised those tests ended up working. That's awesome though.
And I just started with 3.6.3 since it was the latest. Would you like me to add more to `testcase.yml`?

I've never used Vert.x 2.x so I wouldn't be the one to implement those.

wu-sheng commented:
Let's try to add more 3.x version only. Keep 2.x out this is sensible to me.

BFergerson commented:
@wu-sheng, I added all versions from 3.x to the integration test. I can't get any of them to work locally so I'm just going to wait for the SkyWalking Robot to let me know which fail.

wu-sheng commented:
> I added all versions from 3.x to the integration test. I can't get any of them to work locally so I'm just going to wait for the SkyWalking Robot to let me know which fail.

@ascrutae Could you start the tests?

wu-sheng commented:
@BFergerson Now, ApacheCon has opened topic of observability, I think your case(source++) is new about how to use skywalking. if you want to share, welcome to submit cfp here. https://www.apachecon.com/acna19/cfp.html

BFergerson commented:
@wu-sheng, that might actually be pretty interesting. I've never presented at something like ApacheCon before but it's something I'll definitely think about. Are you going?

Also, I'm still getting 100% fails on `vertx-core-3.x-scenario`. I know some of them are failing because of unsupported versions but I can't currently tell with them all failing locally. I'm not able to figure out why they won't work locally :/.

wu-sheng commented:
@BFergerson Look like we only support 3.6+. Is that because of codebase, or something else?

wu-sheng commented:
3.3+, do you want 3.1 and 3.2 supported?

BFergerson commented:
@wu-sheng, I should be able to get it to support 3.2+. This is when Vert.x introduced the `ClusteredMessage` which is what I use to tell if a message on the event bus is going to be written to the wire. Without that object, it looks like determining whether or not a message is going on the wire is rather involved. I don't think I will be able to support 3.0.0 and 3.1.0. Should I still rename the plugin to reflect the versions it supports? If so what would you like me to call it?

wu-sheng commented:
3.x in name should be fine. Can you confirm two thing?
1. This plugin wouldn't make 3.0 and 3.1 codes break 
2. Supported list document include 3.2+

wu-sheng commented on docs/en/setup/service-agent/java-agent/Supported-list.md:
```diff
@@ -63,6 +63,8 @@
 * JSON
   * [GSON](https://github.com/google/gson) 2.8.x (Optional²)
+* Vert.x Ecosystem
+  * Vert.x Core 3.x (Event bus, HTTP client/server)
```
Let's make this more clear.

BFergerson replied:
No problem

wu-sheng commented on apm-sniffer/apm-sdk-plugin/vertx-plugins/vertx-core-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/vertx3/define/HttpClientRequestImplEndInstrumentation.java:
```diff
@@ -0,0 +1,69 @@
+ *
+ * @author brandon.fergerson
+ */
+public class HttpClientRequestImplEndInstrumentation extends ClassInstanceMethodsEnhancePluginDefine {
```
Do we need to add witness class for all instrumentation to lock >= 3.2 and <4.x

BFergerson replied:
I was unaware of witness classes, thanks. I've added them.

BFergerson commented:
@wu-sheng, I think from adding the witness class there should not be any problems with Vert.x 3.0.0 and 3.1.0. I've also updated the documentation to state that for the eventbus only 3.2+ is supported.

I've also split the integration tests so now it tests the web and eventbus separately. The integration tests for the web module are working for all versions in 3.x and eventbus is 3.2+, of course.

wu-sheng commented:
@BFergerson The PR includes a comment issue, causing CI fails, please fix.

wu-sheng commented:
@ascrutae The test results are good enough. You could review codes.

wu-sheng commented:
@BFergerson We finally made this through 28 days.
